romdisk/uart.asm

```diff
@@ -144,48 +144,140 @@ uartsnd_main:
         ; uartrcv main routine.
         ; It currently accepts a single parameter: size to receive.
         ; It will try to open the driver registered as SER0.
-        ; TODO: Implement this command to actually save data in a file.
         ; Parameters:
         ;       HL - ARGV
         ;       BC - ARGC
         ; Returns:
         ;       A - 0 on success
+        ; Modifies:
+        ;       IX, DE
         PUBLIC uartrcv_main
 uartrcv_main:
+        ; Set up stack frame
+        ld ix, 0
+        add ix, sp
+
         dec bc  ; Do not count the command itself
+
+        ; Check that size is present
         ld a, b
         or c
-        ret z
+        jp z, _rcv_usage
+        ; Advance past command 
         inc hl
         inc hl
+
+        ; Load stack frame
+        push hl ; remaining argv (ix-2)
+        push bc ; remaining argc at (ix-4)
+
+        ; Parse the length parameter
         ld a, (hl)
         inc hl
         ld h, (hl)
         ld l, a
-        ; Parse the length parameter
         call parse_int
-        ; Check if A is 0 or not
+        ; Check parse (a should be zero)
         or a
-        ; At the moment, simply return, this command is just for the sake of demonstration
-        ret nz
-        ; Save the buffer size
-        push hl
+        jp nz, _rcv_usage
+        ; Add read/write size to stack frame
+        push hl ; size at (ix-6)
+
+        ; Open UART
         call uart_open
-        ; Save the size in BC
-        pop bc
-        jp m, open_error
-        ; Read from the UART
-        ; DE - Buffer
-        ; BC - Size
-        ; H  - Descriptor
+        jp m, _uart_rcv_clean_and_error
+
+        ; Set up and do read
+        ld b, (ix-5)
```

Comment:
This is very costly in terms of clock cycles and code size byte (4 bytes)
If you can use `pop` instead, it would be smaller and faster

Reply:
See new commit.